File: packages/validation/src/stock-transfer.ts
// =============================================================================
// Stock Transfer (chuyển kho) schemas
// =============================================================================
import { z } from "zod";
import { uuidSchema, listQuerySchema } from "./common";

export const stockTransferStatusSchema = z.enum(["DRAFT", "IN_TRANSIT", "RECEIVED", "CANCELLED"]);
export type StockTransferStatusInput = z.infer<typeof stockTransferStatusSchema>;

export const createStockTransferLineSchema = z.object({
  productId: uuidSchema,
  unitId: uuidSchema,
  fromLocationId: uuidSchema,
  toLocationId: uuidSchema,
  quantity: z.number().positive("Số lượng phải > 0").finite(),
  batchNo: z.string().trim().max(100).optional().nullable(),
  serialNo: z.string().trim().max(100).optional().nullable(),
  expiryDate: z.string().regex(/^\d{4}-\d{2}-\d{2}$/).optional().nullable(),
  notes: z.string().trim().max(1000).optional().nullable(),
  idempotencyKey: uuidSchema,
});

export const createStockTransferSchema = z
  .object({
    fromBranchId: uuidSchema,
    fromWarehouseId: uuidSchema,
    toBranchId: uuidSchema,
    toWarehouseId: uuidSchema,
    transferDate: z.string().regex(/^\d{4}-\d{2}-\d{2}$/),
    expectedReceiptDate: z.string().regex(/^\d{4}-\d{2}-\d{2}$/).optional().nullable(),
    notes: z.string().trim().max(1000).optional().nullable(),
    lines: z.array(createStockTransferLineSchema).min(1, "Phải có ít nhất 1 dòng"),
  })
  .refine((v) => v.fromWarehouseId !== v.toWarehouseId, {
    message: "Kho nguồn và kho đích phải khác nhau",
    path: ["toWarehouseId"],
  })
  .refine(
    (v) => {
      const keys = v.lines.map((l) => l.idempotencyKey);
      return keys.length === new Set(keys).size;
    },
    { message: "Idempotency keys phải unique", path: ["lines"] },
  );

export const updateStockTransferSchema = z.object({
  transferDate: z.string().regex(/^\d{4}-\d{2}-\d{2}$/).optional(),
  expectedReceiptDate: z.string().regex(/^\d{4}-\d{2}-\d{2}$/).optional().nullable(),
  notes: z.string().trim().max(1000).optional().nullable(),
  lines: z.array(createStockTransferLineSchema).optional(),
});

export const receiveStockTransferLineSchema = z.object({
  lineId: uuidSchema,
  receivedQty: z.number().nonnegative("Số lượng nhận phải >= 0").finite(),
});

export const receiveStockTransferSchema = z.object({
  lines: z.array(receiveStockTransferLineSchema).min(1, "Phải có ít nhất 1 dòng nhận"),
  notes: z.string().trim().max(1000).optional().nullable(),
});

export const stockTransferListQuerySchema = listQuerySchema.extend({
  fromBranchId: uuidSchema.optional(),
  toBranchId: uuidSchema.optional(),
  fromWarehouseId: uuidSchema.optional(),
  toWarehouseId: uuidSchema.optional(),
  status: stockTransferStatusSchema.optional(),
  dateFrom: z.string().datetime().optional(),
  dateTo: z.string().datetime().optional(),
});

export type CreateStockTransferInput = z.infer<typeof createStockTransferSchema>;
export type UpdateStockTransferInput = z.infer<typeof updateStockTransferSchema>;
export type ReceiveStockTransferInput = z.infer<typeof receiveStockTransferSchema>;

File: apps/api/src/InventoryPro.Domain/Inventory/StockTransfer.cs
using InventoryPro.Domain.Catalog;
using InventoryPro.Domain.Common;

namespace InventoryPro.Domain.Inventory;

public enum StockTransferStatus
{
    Draft = 0,
    InTransit = 1,
    Received = 2,
    Cancelled = 3,
}

public enum StockTransferLineStatus
{
    Open = 0,
    InTransit = 1,
    Received = 2,
    Cancelled = 3,
}

/// <summary>
/// Phiếu chuyển kho nội bộ. Workflow:
/// DRAFT → IN_TRANSIT (đã xuất khỏi src warehouse) → RECEIVED (đã nhập vào dst warehouse)
/// Có thể CANCELLED từ DRAFT hoặc IN_TRANSIT (tùy nghiệp vụ).
/// Mỗi line khi POST sẽ tạo 2 movements: TRANSFER_OUT (src) + TRANSFER_IN (dst) cùng RefId.
/// Có thể transfer giữa 2 branches khác nhau (inter-branch) hoặc cùng branch.
/// </summary>
public class StockTransfer : TenantScopedEntity
{
    public string TransferNumber { get; set; } = string.Empty;
    public Guid FromBranchId { get; set; }
    public Guid FromWarehouseId { get; set; }
    public Guid ToBranchId { get; set; }
    public Guid ToWarehouseId { get; set; }
    public DateTime TransferDate { get; set; } = DateTime.UtcNow.Date;
    public DateTime? ExpectedReceiptDate { get; set; }
    public string? Notes { get; set; }
    public StockTransferStatus Status { get; set; } = StockTransferStatus.Draft;

    // Out (src) side
    public Guid? OutShippedBy { get; set; }
    public DateTime? OutShippedAt { get; set; }

    // In (dst) side
    public Guid? InReceivedBy { get; set; }
    public DateTime? InReceivedAt { get; set; }

    public string? CancelReason { get; set; }
    public Guid? CancelledBy { get; set; }
    public DateTime? CancelledAt { get; set; }

    public Guid? CreatedBy { get; set; }

    // Navigation
    public ICollection<StockTransferLine> Lines { get; set; } = new List<StockTransferLine>();
}

/// <summary>
/// Dòng chuyển kho. Snapshot product_name/unit_code tại thời điểm tạo.
/// shipped_qty: số lượng đã xuất khỏi src (ban đầu = quantity khi POST).
/// received_qty: số lượng đã nhập vào dst (có thể thiếu hàng → điều chỉnh).
/// </summary>
public class StockTransferLine : TenantScopedEntity
{
    public Guid StockTransferId { get; set; }
    public int LineNo { get; set; }
    public Guid ProductId { get; set; }
    public Guid UnitId { get; set; }
    public string ProductName { get; set; } = string.Empty;
    public string UnitCode { get; set; } = string.Empty;

    // Source side
    public Guid FromLocationId { get; set; }
    public string FromLocationCode { get; set; } = string.Empty;

    // Destination side
    public Guid ToLocationId { get; set; }
    public string ToLocationCode { get; set; } = string.Empty;

    public decimal Quantity { get; set; }
    public decimal ShippedQty { get; set; } = 0;
    public decimal ReceivedQty { get; set; } = 0;

    public string? BatchNo { get; set; }
    public string? SerialNo { get; set; }
    public DateTime? ExpiryDate { get; set; }
    public string? Notes { get; set; }

    public Guid? OutMovementId { get; set; }
    public Guid? InMovementId { get; set; }

    public StockTransferLineStatus Status { get; set; } = StockTransferLineStatus.Open;

    // Navigation
    public StockTransfer? StockTransfer { get; set; }
    public Product? Product { get; set; }
    public UnitOfMeasure? Unit { get; set; }
    public Location? FromLocation { get; set; }
    public Location? ToLocation { get; set; }
}
